### PR TITLE
[WIP][SPARK-47338][SQL] Introduce `UNCLASSIFIED` for default error class

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to introduce `UNCLASSIFIED` for default error class when error class is not defined.

### Why are the changes needed?

In Spark, when an `errorClass` is not explicitly defined for an exception, the method `getErrorClass` returns null so far.

This behavior can lead to ambiguity and makes debugging more challenging by not providing a clear indication that the error class was not set.

### Does this PR introduce _any_ user-facing change?

No API changes, but the user-facing error message will contain `UNCLASSIFIED` when error class is not specified.

### How was this patch tested?

Updated the existing UT (`SparkThrowableSuite`)

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: common/utils/src/main/resources/error/error-classes.json
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:7905]: @itholic Why did you name it with the prefix `_LEGACY_`? Do you plan to eliminate it in the future?

Comment (itholic) [reply common/utils/src/main/resources/error/error-classes.json:7926]: Because I thought that not having an error class assigned basically meant it was a LEGACY error, but I have not very strong opinion. Do you have any preference? Also cc @srielau FYI

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:7926]: > Because I thought that not having an error class assigned basically meant it was a LEGACY error

I would say it is true. `SparkException` can still be raised w/ just a message since it is not fully ported on error classes. For instance:
https://github.com/apache/spark/blob/e170252714e3662c7354321f78a3250114ea7e9e/common/utils/src/main/scala/org/apache/spark/util/SparkThreadUtils.scala#L51-L53

Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:7926]: Since we know the cases when the error class is not set, how about just name the error class like `UNCLASSIFIED`

Comment (itholic) [reply common/utils/src/main/resources/error/error-classes.json:7926]: Sounds reasonable to me. Let me address it.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@itholic Please, update PR's title and description according to your changes.

**Comment by itholic** [comment]
Updated PR title & description. Let me take a look at the CI failure

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by xinrong-meng** [comment]
LGTM once CI passed, thank you!

**Review by xinrong-meng — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (MaxGekk) [inline core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala:617]: how it happens that you have to change this?

File: sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out:685]: The deleted error message seems reasonable. Do you know why it is replaced?

Comment (itholic) [reply sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out:685]: Yeah, I agree that this looks as bit weird.

The reason is that the `UDTF_EXEC_ERROR` is defined from PySpark side, so technically it is `UNCLASSIFIED` from JVM logic as it is not defined from `error-classes.json`.

But the existing error message still shows on user space, such as:

org.apache.spark.SparkException: [UNCLASSIFIED] pyspark.errors.exceptions.base.PySparkRuntimeError: [UDTF_EXEC_ERROR] User defined table function encountered an error in the 'eval' or 'terminate' method: Column 0 within a returned row had a value of None, either directly or within array/struct/map subfields, but the corresponding column type was declared as non-nullable; please update the UDTF to return a non-None value at this location or otherwise declare the column type as nullable.

I'm not sure if it would be better to keep the existing error message for `PythonException` or mark it as `UNCLASSIFIED`.

File: core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala
Comment (itholic) [reply core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala:617]: That is also my question. I believe this error message should not been affected by current change, but CI keep complaining about this.

So I modified it for testing purposes to see if this would really change the response of CI.

**Comment by itholic** [comment]
Let me mark it as a draft for now, as I haven't been able to find a clear cause as to why the CI is complaining.

**Comment by itholic** [comment]
Sorry but let me reopen this PR with current master branch since it's staled too long.

**Pull request closed without merging**
